crates/bevy_ecs/src/event.rs

```diff
@@ -385,57 +360,95 @@ impl<E: Event> ManualEventReader<E> {
     }
 }
 
-trait IteratorExt {
-    fn with_exact_size(self, len: usize) -> ExactSize<Self>
-    where
-        Self: Sized,
-    {
-        ExactSize::new(self, len)
+pub struct ManualEventIterator<'a, E: Event> {
+    iter: ManualEventIteratorWithId<'a, E>,
+}
+
+impl<'a, E: Event> Iterator for ManualEventIterator<'a, E> {
+    type Item = &'a E;
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next().map(|(event, _)| event)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
     }
 }
-impl<I> IteratorExt for I where I: Iterator {}
 
-#[must_use = "iterators are lazy and do nothing unless consumed"]
-#[derive(Clone)]
-struct ExactSize<I> {
-    iter: I,
-    len: usize,
+impl<'a, E: Event> ExactSizeIterator for ManualEventIterator<'a, E> {
+    fn len(&self) -> usize {
+        self.iter.len()
+    }
 }
-impl<I> ExactSize<I> {
-    fn new(iter: I, len: usize) -> Self {
-        ExactSize { iter, len }
+
+impl<'a, E: Event> DoubleEndedIterator for ManualEventIterator<'a, E> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        self.iter.next_back().map(|(event, _)| event)
     }
 }
 
-impl<I: Iterator> Iterator for ExactSize<I> {
-    type Item = I::Item;
+#[derive(Debug)]
+pub struct ManualEventIteratorWithId<'a, E: Event> {
+    reader: &'a mut ManualEventReader<E>,
+    events: &'a Events<E>,
+    range: Range<usize>,
+    marker: PhantomData<E>,
+}
 
-    #[inline]
-    fn next(&mut self) -> Option<I::Item> {
-        self.iter.next().map(|e| {
-            self.len -= 1;
-            e
-        })
+impl<'a, E: Event> ManualEventIteratorWithId<'a, E> {
+    pub fn new(reader: &'a mut ManualEventReader<E>, events: &'a Events<E>) -> Self {
+        // Check if we are behind the oldest event.
+        //
+        // This means we missed events.
+        if reader.last_event_count < events.oldest_id() {
+            reader.last_event_count = events.oldest_id();
+        }
+
+        let range = reader.last_event_count..events.event_count;
+
+        Self {
+            reader,
+            events,
+            range,
+            marker: PhantomData,
+        }
+    }
+
+    /// Iterate over only the events.
+    pub fn without_id(self) -> ManualEventIterator<'a, E> {
+        ManualEventIterator { iter: self }
+    }
+}
+
+impl<'a, E: Event> Iterator for ManualEventIteratorWithId<'a, E> {
+    type Item = (&'a E, EventId<E>);
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.range.next().and_then(|id| self.events.get_event(id)) {
```

Comment:
I do like that this impl is simple and built out of composable pieces, but calling `self.events.get_event(id)` with per-event logic to find the location in the sequence does unnecessary per-event work. The old approach of slicing the "a and b" event queues and chaining them is more efficient. Imo we should port that approach to the ManualEventIteratorWithId impl.

Reply:
This should be the case now